src/DTS/UseCase/AddDocData.php

```diff
@@ -13,8 +13,8 @@ public function __construct(DocDataRepository $repo) {
   }
 
   public function add($type, $key, $data): array {
-    $entity = $this->repo->newDocData($type, $key, $data);
-    $this->repo->save($entity);
-    return $entity->toArray();
+    $docData = $this->repo->newDocData($type, $key, $data);
```

Comment:
I would throw your exceptions from here ...

if they need to be custom exceptions I would make 2 ... 
`InvalidDocTypeException` and `InvalidDataKeyException`

though perhaps it would be better to just have `EmptyFieldException` and pass the field name as an param 

don't collect a list of errors and put them all into an exception .. that is a validator .. and validators expect to find issues so they are not exceptional :P

Reply:
I went back and forth on this yesterday as to where to put the validation. Do I need a validator? Should it be in the usecase? the repository? These are things I considered and tried before settling on what I have. I'll take your suggestions and see what I can do

Reply:
That's also the reason for the UseCase exception in the repository method